torchqdynamics/utils.py

```diff
@@ -179,19 +196,102 @@ def trace(rho: Tensor) -> Tensor:
     return torch.einsum('...ii', rho)
 
 
-def expect(operator: Tensor, state: Tensor) -> Tensor:
-    """Compute the expectation value of an operator on a quantum state or density
-    matrix. The method is batchable over the state, but not over the operator.
+def ptrace(
+    x: Tensor, keep: Union[int, Tuple[int, ...]], dims: Tuple[int, ...]
+) -> Tensor:
+    """Compute the partial trace of a state vector or density matrix.
+
+    Args:
+        x: Tensor of size `(..., n, 1)` or `(..., n, n)`
+        keep: Int or tuple of ints containing the dimensions to keep for the
+            partial trace.
+        dims: Tuple of ints specifying the dimensions of each subsystem in the
+            Hilbert space tensor product.
+
+    Returns:
+        Tensor of size `(..., m, m)` with `m <= n` containing the partially traced out
+            state vector or density matrix.
+
+    Example:
+        >>> rho = tq.tensprod(tq.coherent_dm(20, 2.0),
+                              tq.fock_dm(2, 0),
+                              tq.fock_dm(5, 1))
+        >>> rhoA = tq.ptrace(rho, 0, (20, 2, 5))
+        >>> rhoA.shape
+        torch.Size([20, 20])
+        >>> rhoBC = tq.ptrace(rho, (1, 2), (20, 2, 5))
+        >>> rhoBC.shape
+        torch.Size([10, 10])
+    """
+    # convert keep and dims to tensors
+    if isinstance(keep, int):
+        keep = torch.as_tensor([keep])
+    elif isinstance(keep, tuple):
+        keep = torch.as_tensor(keep)  # e.g. [1, 2]
+    dims = torch.as_tensor(dims)  # e.g. [20, 2, 5]
+    ndims = len(dims)  # e.g. 3
+
+    # check that input dimensions match
+    if not torch.prod(dims) == x.size(-2):
+        raise ValueError(
+            f'Input `dims` {dims.tolist()} does not match the input '
+            f'tensor size of {x.size(-2)}.'
+        )
+    if torch.any(keep < 0) or torch.any(keep > len(dims) - 1):
+        raise ValueError(
+            f'Input `keep` {keep.tolist()} does not match the Hilbert '
+            f'space structure {dims.tolist()}.'
+        )
+
+    # sort keep
+    keep = keep.sort()[0]
+
+    # create einsum alphabet
+    alphabet = list('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ')
+
+    # compute einsum equations
+    eq1 = alphabet[:ndims]  # e.g. 'abc'
+    unused = iter(alphabet[ndims:])
+    eq2 = [next(unused) if i in keep else eq1[i] for i in range(ndims)]  # e.g. 'ade'
+
+    # trace out x over unkept dimensions
+    if is_ket(x):
+        x = x.reshape(-1, *dims)  # e.g. (..., 20, 2, 5)
```

Comment:
Could you use `view` rather than `reshape`? (it's actually the only place in the code where we reshape tensors so we just have to fix it here)